src/Microsoft.ML.Data/Data/DataViewUtils.cs

```diff
@@ -299,214 +294,198 @@ private Splitter(Schema schema)
-                {
-                    return Consolidate(provider, inputs, 128, ref _splitter._consolidateCachePools);
-                }
+            private static RowCursor ConsolidateCore(IChannelProvider provider, RowCursor[] inputs, ref object[] ourPools, IChannel ch)
+            {
+                ch.CheckNonEmpty(inputs, nameof(inputs));
+                if (inputs.Length == 1)
+                    return inputs[0];
+                ch.CheckParam(SameSchemaAndActivity(inputs), nameof(inputs), "Inputs not compatible for consolidation");
+
+                RowCursor cursor = inputs[0];
+                var schema = cursor.Schema;
+                ch.CheckParam(AllCachable(schema, cursor.IsColumnActive), nameof(inputs), "Inputs had some uncachable input columns");
 
-                public static RowCursor Consolidate(IChannelProvider provider, RowCursor[] inputs, int batchSize, ref object[] ourPools)
+                int[] activeToCol;
+                int[] colToActive;
+                Utils.BuildSubsetMaps(schema.Count, cursor.IsColumnActive, out activeToCol, out colToActive);
+
+                // Because the schema of the consolidator is not necessary fixed, we are merely
+                // opportunistic about buffer sharing, from cursoring to cursoring. If we can do
+                // it easily, great, if not, no big deal.
+                if (Utils.Size(ourPools) != schema.Count)
+                    ourPools = new object[schema.Count + (int)ExtraIndex._Lim];
+                // Create the out pipes.
+                OutPipe[] outPipes = new OutPipe[activeToCol.Length + (int)ExtraIndex._Lim];
+                for (int i = 0; i < activeToCol.Length; ++i)
                 {
-                    Contracts.AssertValue(provider);
-                    using (var ch = provider.Start("Consolidate"))
-                    {
-                        return ConsolidateCore(provider, inputs, ref ourPools, ch);
-                    }
+                    int c = activeToCol[i];
+                    ColumnType type = schema[c].Type;
+                    var pool = GetPool(type, ourPools, c);
+                    outPipes[i] = OutPipe.Create(type, pool);
                 }
+                int idIdx = activeToCol.Length + (int)ExtraIndex.Id;
+                outPipes[idIdx] = OutPipe.Create(NumberType.UG, GetPool(NumberType.UG, ourPools, idIdx));
 
-                private static RowCursor ConsolidateCore(IChannelProvider provider, RowCursor[] inputs, ref object[] ourPools, IChannel ch)
+                // Create the structures to synchronize between the workers and the consumer.
+                const int toConsumeBound = 4;
+                var toConsume = new BlockingCollection<Batch>(toConsumeBound);
+                var batchColumnPool = new MadeObjectPool<BatchColumn[]>(() => new BatchColumn[outPipes.Length]);
+                Thread[] workers = new Thread[inputs.Length];
```

Comment:
So threads are used here, can the cursor set be large? Like in the hundreds for example? If so, would tasks be a better alternative?